design/design-doc-stacks.md

```diff
@@ -260,23 +292,178 @@ owners:
 
+### Example `install.yaml`
+
+The `install.yaml` file is expected to conform to a standard Kubernetes YAML file describing a single `Deployment` object.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: crossplane-sample-extension
+  labels:
+    core.crossplane.io/name: "crossplane-sample-extension"
+spec:
+  selector:
+    matchLabels:
+      core.crossplane.io/name: "crossplane-sample-extension"
+  replicas: 1
+  template:
+    metadata:
+      name: sample-extension-controller
+      labels:
+        core.crossplane.io/name: "crossplane-sample-extension"
+    spec:
+      containers:
+      - name: sample-extension-controller
+        image: crossplane/sample-extension:latest
+        env:
+        - name: POD_NAME
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.name
+        - name: POD_NAMESPACE
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
+```
+
+### Example `group.yaml`
+
+```yaml
+group: databases.foocompany.io
+title: Sample Controller Types
+description: |
+    This is a crd group description
+```
+
+### Example `resource.yaml`
+
+```yaml
+resource: mysql
+title: MySQL
+description: |
+  # MySQL Resource by FooCompany
+
+  This is the Crossplane Stack for FooCompany MySQL
+
+  ## Details
+
+  More markdown.
+```
+
+### Example `rbac.yaml`
+
+```yaml
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - secrets
+  - serviceaccounts
+  - events
+  - namespaces
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - patch
+  - delete
+- apiGroups:
+  - samples.crossplane.io
+  resources:
+  - "*"
+  verbs:
+  - "*"
+```
+
+### Example `crd.yaml`
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  labels:
+    controller-tools.k8s.io: "1.0"
+  name: mytypes.samples.crossplane.io
+spec:
+  group: samples.crossplane.io
+  names:
+    kind: Mytype
+    plural: mytypes
+  scope: Namespaced
+  version: v1alpha1
+```
+
+#### Example `crd.yaml` with Stack annotations
+
+It is the job of the SM or a Stack build tool to process the recommended meta-data files into the final CRD installed in the cluster.  These annotations will assist Stack tools in discovery and identification of resources in cluster that can be managed.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: mytypes.samples.crossplane.io
+  labels:
+    controller-tools.k8s.io: "1.0"
+  annotations:
+    stack.stacks.crossplane.io/name: "crossplane-sample-extension"
+    stack.stacks.crossplane.io/group-title: "Title of the Group"
+    stack.stacks.crossplane.io/group-description: |
+        Description of the Group
+    stack.stacks.crossplane.io/resource-title: "Title of the Resource"
+    stack.stacks.crossplane.io/resource-description: |
+        Description of the Resource
+    stack.stacks.crossplane.io/icon-uri: data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg"/>
```

Comment:
I thought we landed on no URIs for the image data and that the image data was to be base64 encoded instead?

Reply:
The Extension CRD is still using the well-defined `{icons: [{base64data, mediatype}]}` format.

We are using annotations to propagate stack metadata into the leaf CRDs because we don't want to force a `spec` convention on new controllers or become incompatible with existing controllers that wish to be (re)packaged as stacks.
 
The annotations of the leaf CRDs can not be formally structured.  These annotation must be either a single line string using a well recognized format that encapsulates the icon properties or they can be a multi-line string using a structuring format like YAML or JSON that recites or translates the original icon schema format for the base Extension icon or specific leaf icons.

```yaml
annotations:
   stack.stacks.crossplane.io/icon: >
     base64data: iVBORw0KGgoAAAANSUhEUgAAAOEAAADZCAYAAADWmle6AAAACXBIWXMA
     mediatype: image/png
```

```yaml
annotations:
    stack.stacks.crossplane.io/icon-uri: data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg==
```

```yaml
annotations:
    stack.stacks.crossplane.io/icon-uri: data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg"/>
```

I would also venture that single string annotations are more usable from CLI tools and other consumers.

While translating the `resource.yaml` file to propose annotations, I wanted to avoid the duplicative elements of the file.  The resource.yaml file makes reference to the CRD name itself, for example.  This naturally lead to a flattening of the elements into individual annotations representing each property of the `resource.yaml`.

```yaml
annotations:
  stack.stacks.crossplane.io/resource-title: foo
  stack.stacks.crossplane.io/resource-description: bar
```

```yaml
meta:
  name: thistype
annotations:
  stack.stacks.crossplane.io/resource: |
    resource: thistype
    title: foo
    description: bar
```

To avoid complexity, it follows that the icon structure should be also be flattened if the `resource.yaml` structure is flattened.

I realize that I have am not attempting to flatten the ui-yaml which seems hypocritical but we've expressly described the ui-yaml as having an undefined format.

For icons, I'm not pushing for an all purpose URI ([RFC3986](https://tools.ietf.org/html/rfc3986)), rather I think Data URLs are sufficient to express our requirements in a single line string [RFC2397](https://tools.ietf.org/html/rfc2397)) and avoid the complications of [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#What_requests_use_CORS) and [CSRF](https://en.wikipedia.org/wiki/Cross-site_request_forgery#Example_and_characteristics).

Reply:
I would also like to know if `stack.stacks.crossplane.io/` or   `stacks.crossplane.io/` is the preferred annotation prefix.

Reply:
Handling the encoding parameter of the Data URL may be cumbersome to implement if the image must be processed (and not just handed off to a browser).

I could add that the Data URLs SHOULD or MUST be base64 encoding to avoid that (or use the `{icon:{base64data,mimetype}}` format).

Reply:
Thanks for the thoughtful response @displague.  I don't think I understood the "data URL" concept until you explained it now and pointed at [RFC 2397](https://tools.ietf.org/html/rfc2397).  TIL 🤓 That format seems reasonable since it should be very easy for a browser to display as is with no translation logic by a UI.

What do you think about the following?
* name the annotation field `stack.stacks.crossplane.io/icon-data-uri` to make it more intuitive that it's intended to be in a data URI format.
* add a note about the format of this field and maybe link to the RFC for posterity.
* for consistency, should the `Extension` CR's icon fields also be in this data URI format?

I am good with flattening the resource and group information into specific annotation fields (e.g. `stacks.crossplane.io/group-title`).

I think `stacks.crossplane.io/` makes sense because that fully identifies the group that all these annotations would belong to, similar to how `app.kubernetes.io/` is used [here](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels).

Let me know if you want to do a quick synchronous meetup to drive this to completion, thanks for your efforts!

Reply:
> name the annotation field `stack.stacks.crossplane.io/icon-data-uri` to make it more intuitive that it's intended to be in a data URI format.
That is certainly the intent today, but I don't know that this should always be the case.  I'll go with the flow and name it `icon-data-uri`.  I shared this thought when naming it.  We can always deprecate `data` and prefer `icon-uri` later if we open up to more schemes like `https`.

I don't have a strong view on which format is best in the Extension record.  I suspect we may want to add more properties there beyond what the data URI can convey.  We could use query parameters, but no tool is going to know what to do with those.

```yaml
kind: Extension
spec:
 icons:
 - base64data: iVBORw0KGgoAAAANSUhEUgAAAOEAAADZCAYAAADWmle6AAAACXBIWXMA
   mediatype: image/png
```

We lose the ability to add more properties If we go from that to this..

```yaml
kind: Extension
spec:
 icons:
 - data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAOEAAADZCAYAAADWmle6AAAACXBIWXMA
```

And we aren't winning any pageants with a single property map either:

```yaml
- datauri: data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAOEAAADZCAYAAADWmle6AAAACXBIWXMA
- datauri: data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAOEAAADZCAYAAADWmle6AAAACXBIWXMA
- datauri: data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAOEAAADZCAYAAADWmle6AAAACXBIWXMA
```

Incase we ever want to sneak more properties in there, like an icon title, ID, or usage hints, I think we'll benefit from the expanded format currently found in the Extension record.

Reply:
I'm marking this as resolved, @jbw976. Your changes have been incorporated or addressed verbally above.